### PR TITLE
Randomized expiration times for cache entries

### PR DESCRIPTION
**Pull request by m-ashour**
Signed-off-by: Maha Ashour <mashour@redhat.com>

### CODE REVIEW

**Comment by m-ashour** [comment]
These are minor changes but I think they can cause some not-minor issues:

If we set a random expiration time (in hours) anywhere from 1 hour to 6 months (about 4320 hrs), we still have a 50% chance of time slot collision with just 77 cache entries. With 100,000 random time slots that number only goes up to 372 entries. And given that we have thousands of cache entries, I'm not sure what impact this will have on hitting the quota.
I used the [birthday paradox](https://en.wikipedia.org/wiki/Birthday_problem) calculation for this: sqrt(2*4320*ln(1/(1-0.5))) = 77
For reference, the maximum expiration time in Redis is 2 147 483 647 seconds (which is about 68 years).

mirror_requests.py will check if an item is in cache before it 'gets' it. It is very possible that the cache entry will expire after the 'exists' check and before the lookup. That would raise a KeyError. So I think we'd require a try/catch block in addition to (or instead of) checking whether the key exists.

**Comment by apahim** [comment]
1. Can we refresh the expires on read? That would avoid us of overengineering the expiration time randomness.
2. I'd use the try/except instead of the check. That, as you said, would make the operation atomic.

File: ghmirror/core/mirror_requests.py [hunk 1]
```diff
@@ -105,8 +105,8 @@ def online_request(method, url, auth, data=None):
     resp.headers['X-Cache'] = 'ONLINE_MISS'
     # Caching only makes sense when at least one
     # of those headers is present
-    if any(['ETag' in resp.headers,
-            'Last-Modified' in resp.headers]):
+    if resp.status_code == 200 and any(['ETag' in resp.headers,
+                                        'Last-Modified' in resp.headers]):
```
Comment (apahim) [inline ghmirror/core/mirror_requests.py:109]: this change belongs to a different commit...

File: ghmirror/data_structures/redis_data_structures.py [hunk 2]
```diff
@@ -50,7 +51,9 @@ def __getitem__(self, item):
     def __setitem__(self, key, value):
         sr_key = self._serialize(key)
         sr_value = self._serialize(value)
-        self.wr_cache.set(sr_key, sr_value)
+        rand_val = randint(1, 4320)
+        # set entry to expire in rand_val hrs
```
Comment (apahim) [inline ghmirror/data_structures/redis_data_structures.py:55]: it would be more useful for the comment to explain the 4320 above instead of just reading the code below

**Comment by m-ashour** [comment]
> 1. Can we refresh the expires on read? That would avoid us of overengineering the expiration time randomness.
> 2. I'd use the try/except instead of the check. That, as you said, would make the operation atomic.

there is no atomic GETEX so we have a few options to do this: 

a) on every read, first check if the item EXISTS, then refresh expiration with EXPIRE (if it happens to expire in that moment the refresh value will be ignored but it will be done on the next SET) and then GET the value. This would necessitate 3 RT to Redis. We were already doing 2 RTs unnecessarily (since the GET returns a special value nil if the item is not in the cache).

b) we can do some fun stuff - Redis [transactions](https://redis.io/topics/transactions). It is a way to group several commands and execute them atomically and in order as one "transaction". In the python module this is implemented as a kind of optimistic concurrency control using [pipelines](https://github.com/andymccurdy/redis-py#pipelines) and watches. This may slow things down a bit, but I'd guess that it's still faster than 3 RTs.

File: ghmirror/data_structures/redis_data_structures.py (hunk 2 above)
Comment (m-ashour) [reply ghmirror/data_structures/redis_data_structures.py:55]: ha you're right. i had that in before and accidentally deleted it

File: ghmirror/core/mirror_requests.py (hunk 1 above)
Comment (m-ashour) [reply ghmirror/core/mirror_requests.py:109]: its in a separate commit now. want it in a separate PR?

Comment (apahim) [reply ghmirror/core/mirror_requests.py:109]: No, looks good, thanks!
